### PR TITLE
Support login to UAA using API token

### PR DESCRIPTION
**Pull request by vuil**
### What this PR does / why we need it

Enables login via a previously obtained API token to a UAA based endpoint.
Also: fixed a bug where the expiration time on a token refresh is incorrect set to too close to current time.

Note : adding more tests and waiting for the alternative client id to be determined.

### Which issue(s) this PR fixes

Fixes #

### Describe testing done for PR

Obtain refresh token from CLI context
TANZU_API_TOKEN=_obtainedtoken_  tanzu login --endpoint uaa-based-endpoint

Updated and ran unit tests.

### Release note

```release-note
Token obtained from the "api-token create" command can be used as value to environment variable TANZU_API_TOKEN to connect to uaa-based endpoint.
```

### CODE REVIEW

File: pkg/command/context.go [hunk 1]
```diff
@@ -732,11 +732,53 @@ func getSelfManagedOrg(c *configtypes.Context) (string, string) {
...
+		return nil, errors.New("auth endpoint not set")
+	}
 	uaaEndpoint := c.AdditionalMetadata[config.TanzuAuthEndpointKey].(string)
 	log.V(7).Infof("Login to UAA endpoint: %s", uaaEndpoint)
 
-	claims, err := doInteractiveLoginAndUpdateContext(c, uaaEndpoint)
+	apiTokenValue, ok := os.LookupEnv(config.EnvAPITokenKey)
+	// Use API Token login flow if TANZU_API_TOKEN environment variable is set, else fall back to interactive login flow
+	if ok {
+		log.Info("API token env var is set")
```
Comment (anujc25) [inline pkg/command/context.go:773]: nit(existing): Should we just mention `TANZU_API_TOKEN` instead of mentioning `API token` here?

**Review by anujc25 — APPROVED** [review]
LGTM. Thanks.

File: pkg/auth/common/token.go [hunk 2]
```diff
@@ -20,6 +20,8 @@ const (
 	extraIDToken = "id_token"
 )
 
+var currentTime = time.Now
```
[the remark this answers is not in the document]
Comment (marckhouzam) [reply pkg/auth/common/token.go:23]: Oh, I see it is for testing. ok, looks good.

File: pkg/auth/common/token.go [hunk 3]
```diff
@@ -171,7 +173,7 @@ func ParseToken(tkn *oauth2.Token, idpType config.IdpType) (*Claims, error) {
 func IsExpired(tokenExpiry time.Time) bool {
 	// refresh at half token life
 	two := 2
-	now := time.Now().Unix()
+	now := currentTime().Unix()
```
[the remark this answers is not in the document]
Comment (marckhouzam) [reply pkg/auth/common/token.go:176]: For testing.

File: pkg/auth/common/token.go (hunk 2 above)
Comment (marckhouzam) [inline pkg/auth/common/token.go:23]: Would you mind adding a comment to explain that `currentTime` is used to allow tests to override it?

File: pkg/auth/common/login_handler.go [hunk 4]
```diff
@@ -122,6 +122,18 @@ func WithCertInfo(tlsSkipVerify bool, caCertData string) LoginOption {
 	}
 }
 
+// WithClientID specifies a OAuth Client ID to use
+func WithClientID(clientID string) LoginOption {
+	return func(h *TanzuLoginHandler) error {
+		fmt.Printf("update h.clientID, h.oauthclientid = %s\n", clientID)
```
[the remark this answers is not in the document]
Comment (vuil) [reply pkg/auth/common/login_handler.go:128]: removed

[the remark this answers is not in the document]
Comment (vuil) [reply pkg/auth/common/login_handler.go:128]: Stray, removed.

File: pkg/auth/uaa/tanzu.go [hunk 5]
```diff
@@ -4,20 +4,28 @@
...
 	// only allowed Redirect URI and does not have an associated client secret.
-	tanzuCLIClientID     = "tp_cli_app"
-	tanzuCLIClientSecret = ""
-	defaultListenAddress = "127.0.0.1:0"
-	defaultCallbackPath  = "/callback"
+	tanzuCLIClientID = "tp_cli_app"
+	// Alternate client ID for UAA associated with a longer refresh token
+	// validity. Use this for CLI use cases where it is impractical to
+	// interactively reauthenticate once the refresh token expires.
+	tanzuCLIClientIDExtended = "tp_cli_app_ext"
```
Comment (marckhouzam) [inline pkg/auth/uaa/tanzu.go:25]: Do we need to set this to the same as the other one for now?

**Review by marckhouzam — APPROVED** [review]
Looks good (I didn't try it).
One question about the extendedApp value.

File: pkg/auth/uaa/tanzu.go (hunk 5 above)
Comment (marckhouzam) [reply pkg/auth/uaa/tanzu.go:25]: Ok, I saw on chat that you will leave the logic to fallback to the `TANZU_CLI_USE_ALTERNATE_UAA_CLIENT` variable but that we won't use it for now

File: pkg/auth/common/token.go (hunk 2 above)
Comment (anujc25) [reply pkg/auth/common/token.go:23]: As will be removed and addressed as part of https://github.com/vmware-tanzu/tanzu-cli/pull/820#discussion_r1790816475 by using separate package.

**Pull request merged**